### DISCUSSION TITLE
Child result set codegen, column always named `child1`?

### DISCUSSION THREAD

**Discussion by amyworrall** [ricomariani/CG-SQL-author#104] [Q&A] [answered]
When I'm doing something like this, is there a way to get the child result set column named something else?

```
create proc PerspectiveContent()
begin
-- This is syntactic sugar to do the parent/child join, but AFAIK it always names
-- the child result set column as `child1`
out union call PerspectiveContentGroup() join call PerspectiveContentItem() using (roomID);
end;
```

**Comment by ricomariani** [comment]
No the child names are synthetic, the are always childN it's hard coded.  There is no good reason for this other than that there could be arbitarily many of them and it doesn't know what to call them.  I can't use the proc name because it could be shared.  I guess we could add an attribute or something to give it a hint?


________________________________
From: amyworrall ***@***.***>
Sent: Sunday, September 22, 2024 4:54 PM
To: ricomariani/CG-SQL-author ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [ricomariani/CG-SQL-author] Child result set codegen, column always named `child1`? (Discussion #104)


When I'm doing something like this, is there a way to get the child result set column named something else?

create proc PerspectiveContent()
begin
-- This is syntactic sugar to do the parent/child join, but AFAIK it always names
-- the child result set column as `child1`
out union call PerspectiveContentGroup() join call PerspectiveContentItem() using (roomID);
end;

**Comment by amyworrall** [comment]
An attribute would be fine. Or I was wondering about allowing an `AS name` inside the `OUT UNION` line?

**Comment by ricomariani** [comment]
It would be novel syntax because there can be any number of children.

As it is I have to think about how this generalizes.

The annoying thing is that the most common case is one child you just want to name and we have to complicate it on the off chance it is weird...
________________________________

**Comment by ricomariani** [comment]
Open an issue 😄
________________________________

**Comment by amyworrall** [comment]
One further question, is there a way to do the join on differently named fields? For example if the join field in the parent was named `roomID` but in the child sproc was just named `id`?

**Comment by ricomariani** [comment]
The sugar assumes the fields match.

Of course there is no law that says you have to use the sugar...

________________________________

**Comment by amyworrall** [comment]
this is true, but I like the sugar otherwise!

BTW, I'll be giving a conference talk about how I'm building my codegen on top of CGSQL, at Pragma Conf in October. Not sure if the talks are recorded but thought you might like to know :)

**Comment by ricomariani** [comment]
Cool.

I'll see what I can do about hinting the sugar
________________________________

**Comment by ricomariani** [comment, marked as answer]
It turns out there is already syntax for this:


proc test_parent_child()
begin
  out union
   call test_parent(2) join call test_child(1) using (x) as my_child;
end;

You can name each child result if you want to otherwise you get child1...n
________________________________

**Comment by amyworrall** [comment]
Hi @ricomariani , just wondering if you can do a parent/child join inside a shared fragment?

I'd like a shared fragment that returns rows where one column is a child result set. Is this possible?

**Comment by ricomariani** [comment]
No but you could make a macro for that kind of thing.

The shared fragment basically becomes a CTE so whatever you can do in a CTE.  It takes two queries (or  more) to make parent child).  You can do recursive queries with a single with recursive but you'd get something that is shaped like a rectangle still.
________________________________
